src/data_store/handler.clj

```diff
@@ -1,35 +1,75 @@
     (cond
-      (nil? command)
-      [store "-Error no command \r\n"]
-      (= command "PING")
-      [store "$4\r\nPONG\r\n"]
-      (= command "ECHO")
-      (if (nth parsed 4 false)
-        [store (str "+" (nth parsed 4) delim)]
-        [store "-Error nothing to echo\r\n"])
-      (= command "SET")
-      (cond
-        (not (nth parsed 6 nil))
-        [store "-Error nothing to set\r\n"]
-        (not (string? (nth parsed 6 nil)))
-        [store "-Error not a string\r\n"]
-        :else [(into store {(nth parsed 4) (nth parsed 6)}) "+OK\r\n"])
-      (= command "GET")
-      (cond
-        (not (nth parsed 4 nil))
-        [store "-Error nothing to get\r\n"]
-        (not (contains? store (nth parsed 4)))
-        [store "$-1\r\n"]
-        (not (string? (get store (nth parsed 4 nil))))
-        [store "-Error not a string\r\n"]
-        :else [store (str "+" (get store (nth parsed 4)) delim)])
-      :else [store "-Error invalid command\r\n"])))
+      (or (not k) (not v))
+      [store (str "-Error nothing to set" delim)]
+      (not (string? v))
+      [store (str "-Error not a string" delim)]
+      :else (let [sub-command (nth input 3 nil)
+                  sub-command-arg (nth input 4 nil)]
+              (cond
+                (and (= sub-command "EX") sub-command-arg)
+                [(into store {k {:val v :exp (+ time (* 1000 (Integer/parseInt sub-command-arg)))}}) (str "+OK" delim)]
+                (and (= sub-command "PEX") sub-command-arg)
+                [(into store {k {:val v :exp (+ time (Integer/parseInt sub-command-arg))}}) (str "+OK" delim)]
+                (and (= sub-command "EXAT") sub-command-arg)
+                [(into store {k {:val v :exp (* 1000 (Integer/parseInt sub-command-arg))}}) (str "+OK" delim)]
+                (and (= sub-command "PXAT") sub-command-arg)
+                [(into store {k {:val v :exp (Integer/parseInt sub-command-arg)}}) (str "+OK" delim)]
+                :else [(into store {k {:val v}}) (str "+OK" delim)])))))
+
+(defn command-expire [input store time]
+  (let [store-key (nth input 1 nil)
+        exp-time (nth input 2 nil)]
+    (if (and
+         store-key
+         exp-time
+         (contains? store store-key))
+      [(into store {store-key (into (get store store-key) {:exp (+ time (* 1000 (Integer/parseInt exp-time)))})})
+       (str ":1" delim)]
+      [store (str ":0" delim)])))
+
+(defn command-echo [input store _]
+  (if-let [msg (second input)]
+    [store (str "+" msg delim)]
+    [store (str "-Error nothing to echo" delim)]))
+
+(defn command-ping [input store _]
+  (if-let [msg (nth input 2 nil)]
+    [store (str "+" "PONG" " " msg delim)]
+    [store (str "$4" delim "PONG" delim)]))
+
+(defn command-unknown [_ store _]
+  [store (str "+OK" delim)])
+
+(defn command-handler [command & args]
+  (if-let [dispatch-handler (ns-resolve 'data-store.handler (symbol (str "command-" command)))]
+    (apply dispatch-handler args)
+    (apply command-unknown args)))
```

Comment:
Gets the command handler based on the name of the function.

If the command is `ping`, the function will be `command-ping`.